src/checkpoints/checkpoints.cpp

```diff
@@ -69,92 +71,211 @@ namespace cryptonote
         END_KV_SERIALIZE_MAP()
   };
 
-  //---------------------------------------------------------------------------
-  checkpoints::checkpoints()
-  {
-  }
   //---------------------------------------------------------------------------
   bool checkpoints::add_checkpoint(uint64_t height, const std::string& hash_str)
   {
     crypto::hash h = crypto::null_hash;
-    bool r = epee::string_tools::hex_to_pod(hash_str, h);
+    bool r         = epee::string_tools::hex_to_pod(hash_str, h);
     CHECK_AND_ASSERT_MES(r, false, "Failed to parse checkpoint hash string into binary representation!");
 
-    // return false if adding at a height we already have AND the hash is different
-    if (m_points.count(height))
+    CRITICAL_REGION_LOCAL(m_lock);
+    if (m_points.count(height)) // return false if adding at a height we already have AND the hash is different
+    {
+      checkpoint_t const &checkpoint = m_points[height];
+      crypto::hash const &curr_hash  = checkpoint.block_hash;
+      CHECK_AND_ASSERT_MES(h == curr_hash, false, "Checkpoint at given height already exists, and hash for new checkpoint was different!");
+    }
+    else
     {
-      CHECK_AND_ASSERT_MES(h == m_points[height], false, "Checkpoint at given height already exists, and hash for new checkpoint was different!");
+      checkpoint_t checkpoint = {};
+      checkpoint.type         = checkpoint_type::predefined_or_dns;
+      checkpoint.block_hash   = h;
+      m_points[height]        = checkpoint;
     }
-    m_points[height] = h;
+
     return true;
   }
   //---------------------------------------------------------------------------
-  bool checkpoints::is_in_checkpoint_zone(uint64_t height) const
+  static bool add_vote_if_unique(checkpoint_t &checkpoint, service_nodes::checkpoint_vote const &vote)
   {
-    return !m_points.empty() && (height <= (--m_points.end())->first);
+    CHECK_AND_ASSERT_MES(checkpoint.block_hash == vote.block_hash,
+                         false,
+                         "Developer error: Add vote if unique should only be called when the vote hash and checkpoint hash match");
+
+    // TODO(doyle): Factor this out, a similar function exists in service node deregister
+    CHECK_AND_ASSERT_MES(vote.voters_quorum_index < service_nodes::QUORUM_SIZE,
+                         false,
+                         "Vote is indexing out of bounds");
+
+    const auto compare_vote_and_voter_to_signature = [](service_nodes::checkpoint_vote const &a, service_nodes::voter_to_signature const &b) {
+      return a.voters_quorum_index < b.quorum_index;
+    };
+
+    auto signature_it = std::upper_bound(checkpoint.signatures.begin(), checkpoint.signatures.end(), vote, compare_vote_and_voter_to_signature);
```

Comment:
As discussed in the office, this might insert duplicate votes.

Reply:
Updated. They don't need to be sorted so can just search linearly with 10 elements won't matter too much. Might sort them at the end when entering into the DB so the data is nicely formed.